### DISCUSSION TITLE
Scope definition in multi module project

### DISCUSSION THREAD

**Discussion by roman-stetsenko-reed** [hmlongco/Factory#62] [Q&A] [answered]
I have two modules: **AccountLoading**
```
public extension Container {
    var accountLoader: Factory<AccountLoading?> { Factory(self, { nil }) }
}

public struct Account {
    public let name: String
    public init(name: String) {
        self.name = name
    }
}

public protocol AccountLoading {
    func load() -> Account
}
```

and **AccountInfo**
```
public class AccountInfo: ObservableObject {
    @Injected(\.accountLoader) var loader
    @Published public var account: Account = Account(name: "")
    public init() {}
    public func load() {
        guard let loader else { return }
        account = loader.load()
    }
}
```

In third module I define dependency for Account info:
```
class ApiAccountLoading: AccountLoading {
    func load() -> Account { Account(name: "API") }
}

extension Container: AutoRegistering {
    public func autoRegister() {
        accountLoader.register { ApiAccountLoading() }
    }
}

let accountInfo = AccountInfo()
```

With current implementation scope for `Container.accountLoader` is defined in **AccountLoading** module and is `nil` in this example.
But I want to be able to define scope in `autoRegister` function. How I can do this? Because right now 
_"'scope' is inaccessible due to 'internal' protection level"_

**Comment by hmlongco** [comment, marked as answer]
Umm... I'd have to think about the ramifications of allowing the user to change scope after the fact.

What's the drawback to having the publicly visible Factory define the scope? Then everyone who sees that definition has a better understanding of what to expect.

**Reply by roman-stetsenko-reed** [reply]
Right now `Container` extension in **AccountLoading** defines the scope for all its usages. And I don't have flexibility in third module to decide if I want it to be `singleton` or `unique`.
For me it looks more logical to define scope during dependencies assembly and **AccountLoading** shouldn't know anything about how it will be used.

**Reply by hmlongco** [reply]
I'll have to look at how that affects the architecture. Once that Factory is defined within the computed variable its parameters are set. Registering a new factory basically generates an entry in a lookup table in the ContainerManager and if I wanted to do support changing scope after the fact I'd have to do something similar.

**Reply by hmlongco** [reply]
I agree and it turns that there was a relatively simple solution to the problem. Pull the current develop branch.
```swift
extension Container: AutoRegistering {
    public func autoRegister() {
        externalType.register(scope: .singleton) { MyExternalType() }
    }
}
```

**Reply by roman-stetsenko-reed** [reply]
Thanks for helping with this so fast!

**Comment by hmlongco** [comment]
BTW, noticed your code sample preferred the formal Factory definition.
```swift
public extension Container {
    var accountLoader: Factory<AccountLoading?> { Factory(self, { nil }) }
}
```
Don't like the convenience functions?

**Reply by roman-stetsenko-reed** [reply]
Don't pay much attention, as I just quickly prototyped this sample. Inline functions for scope looks nice, maybe a little bit unusual.

**Reply by hmlongco** [reply]
It is, but its the best I have until Swift figures out a way to add variables using an extension...
